src/ReceiverLink.cs

```diff
@@ -97,29 +106,62 @@ public void Start(int credit, MessageCallback onMessage = null)
+        /// number of messages in flight. Will not return negative credit.
+        /// </summary>
+        /// <param name="count_rcv">reciever delivery count</param>
+        /// <param name="count_snd">sender deliver count</param>
+        /// <param name="credit">desirec message credit</param>
+        /// <returns></returns>
+        internal int ComputeCredit(int count_rcv, int count_snd, int credit)
+        {
+            uint rCredit = unchecked((uint)count_rcv + (uint)credit - (uint)count_snd);
+            int iCredit = Math.Max((int)rCredit, 0);
+            return iCredit;
+        }
+
+        /// <summary>
+        /// Compute receive delivery count at which autoRestore will
+        /// issue a new credit flow.
+        /// </summary>
+        /// <param name="count_rcv">receiver delivery count</param>
+        /// <param name="credit">credit setting</param>
+        /// <returns></returns>
+        internal int ComputeRestoreCount(int count_rcv, int credit)
+        {
+            int rCount = unchecked(count_rcv + ((credit + 1) / 2));
+            return rCount;
+        }
+
         /// <summary>
         /// Sets a credit on the link. A flow is sent to the peer to update link flow control state.
+        /// Lowering the credit value may result in send a flow with zero credits.
+        /// The autoRestore receive delivery count is computed.
         /// </summary>
         /// <param name="credit">The new link credit.</param>
-        /// <param name="autoRestore">If true, link credit is auto-restored when a message is accepted
-        /// or rejected by the caller. If false, caller is responsible for manage link credits.</param>
+        /// <param name="autoRestore"> If false, caller is responsible for
+        /// managing link credit. If true, link credit is auto-restored when
+        /// a message is disposed (Accept, Reject, Modified, Released) by the caller
+        /// and the receiver has disposed of enough messages so that a new credit of at least
+        /// credit/2 may be issued.</param>
         public void SetCredit(int credit, bool autoRestore = true)
         {
-            uint dc;
+            int flowCredit;
             lock (this.ThisLock)
             {
                 if (this.IsDetaching)
                 {
                     return;
                 }
 
-                this.totalCredit = autoRestore ? credit : 0;
                 this.credit = credit;
-                this.restored = 0;
-                dc = this.deliveryCount;
+                this.autoRestore = autoRestore;
+                flowCredit  = ComputeCredit(deliveryCountRcv, deliveryCountSnd, credit);
+                restoreCountRcv = ComputeRestoreCount(deliveryCountRcv, credit);
+                deliveryLimitSnd = unchecked(deliveryCountRcv + credit);
             }
-
-            this.SendFlow(dc, (uint)credit, false);
+            this.SendFlow((uint)this.deliveryCountRcv, (uint)flowCredit, false);
```

Comment:
The delivery count value used in the flow should probably be taken once under the lock as before and then reused in the subsequent calculations and flow, since it needs to align exactly with the derived flowCredit value. (EDIT: similar to subsequent comment later for DisposeMessage)

Like in DisposeMessage, it also seems that it should be deliveryCountSnd that the flow delivery count is based on here, since additional deliveries could have already arrived (essentially what delivery-count in flow exists to convey, as a reference point) before the calculations and so that is already incorporated in the new flowCredit value calculated here. Using deliveryCountRcv as the base point would effectively undergrant credit to whatever extent it had already been used beyond the restore point.

Reply:
Agreed on both points.